test/controllers/products_controller_test.rb

```diff
@@ -79,32 +68,129 @@
         must_respond_with :redirect
         must_redirect_to product_path(new_product_id)
       end
+
+      it "renders bad_request and does not update the DB for bogus data" do
+        merchant = perform_login
+        bad_prod_name = {product: {name: nil, price: 1000}}
+
+        expect {
+          post products_path, params: bad_prod_name
+        }.wont_change "Product.count"
+
+        must_respond_with :bad_request
+        expect(flash[:error]).must_include "Could not add new product"
+        expect(flash[:name]).must_include "can't be blank"
+      end
+
+      describe "logged out user" do
+        it "cannot create a product with valid data" do
+          new_product = {product: {name: "Something amazing", price: 1000}}
+          expect {
+            post products_path, params: new_product
+          }.wont_change "Product.count"
+
+          must_respond_with :bad_request
+          expect(flash[:error]).must_equal "You must be logged in to create a new product"
+        end
+      end
     end
+  end
+
+  describe "edit" do
+    describe "logged in merchant" do
+      it "succeeds" do
+        merchant = perform_login
+
+        get edit_product_path(merchant.products.first.id)
+
+        must_respond_with :success
+      end
+
+      it "will redirect if given an invalid product id" do
+        merchant = perform_login
 
-    it "renders bad_request and does not update the DB for bogus data" do
-      merchant = perform_login
-      bad_prod_name = { product: { name: nil, price: 1000 } }
+        get edit_product_path(-1)
 
-      expect {
-        post products_path, params: bad_prod_name
-      }.wont_change "Product.count"
+        must_respond_with :redirect
+      end
+    end
+    describe "not logged in" do
+      it "redirects user" do
+        merchant = merchants(:merchant_1)
+        # do not do login
+        get edit_product_path(merchant.products.first.id)
 
-      must_respond_with :bad_request
-      expect(flash[:error]).must_include "Could not add new product"
-      expect(flash[:name]).must_include "can't be blank"
+        # must_respond_with :redirect
+        expect(flash[:error]).must_equal "You must be logged in to edit a product"
```

Comment:
is this meant to be commented out?

Reply:
also if it does redirect, where does it redirect to?

Reply:
not meant to be commented out, added a redirect to products_path :)